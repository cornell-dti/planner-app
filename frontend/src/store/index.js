// @flow strict

import { createStore } from 'redux';
import type { Store } from 'redux';
import rootReducer from './reducers';
import type { State } from './store-types';
import type { Action } from './action-types';

const store: Store<State, $Subtype<Action>> = createStore(rootReducer);
export default store;
