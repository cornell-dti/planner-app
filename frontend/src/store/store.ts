import { createStore, Store } from 'redux';
import rootReducer from './reducers';
import { State } from './store-types';
import { Action } from './action-types';
import { initialStateForTesting } from './state';

export type GlobalStore = Store<State, Action>;

/**
 * The redux store. It should only be given to the react-redux provider.
 */
export const store: GlobalStore = createStore(rootReducer);

/**
 * The redux store for testing.
 */
export const storeForTesting: GlobalStore = createStore(
  (initialState: State = initialStateForTesting, action: Action): State => rootReducer(
    initialState, action,
  ),
);

/**
 * Dispatch an action and returns the same dispatched action.
 */
export const dispatchAction = (action: Action): Action => store.dispatch(action);
