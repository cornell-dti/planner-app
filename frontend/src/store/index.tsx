import React, { ReactElement } from 'react';
import { Provider as ReactReduxProvider } from 'react-redux';
import { store, storeForTesting } from './store';

type Props = { readonly children: ReactElement };

export const Provider = ({ children }: Props): ReactElement => (
  <ReactReduxProvider store={store}>{children}</ReactReduxProvider>
);

export const ProviderForTesting = ({ children }: Props): ReactElement => (
  <ReactReduxProvider store={storeForTesting}>{children}</ReactReduxProvider>
);
