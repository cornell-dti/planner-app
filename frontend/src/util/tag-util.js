// @flow strict

import type { ComponentType } from 'react';
import type { Tag } from '../store/store-types';
import { error } from './general-util';
import { stateConnect } from '../store/react-redux-util';
import type { ConnectedComponent } from '../store/react-redux-util';

/**
 * ID of the none tag.
 * @type {number}
 */
export const NONE_TAG_ID = -1;

/**
 * The none tag.
 * @type {Tag}
 */
export const NONE_TAG: Tag = {
  id: NONE_TAG_ID, name: 'None', color: 'gray', classId: null,
};

/**
 * Returns the tag from a list of tags given the name.
 *
 * @param {Tag[]} tags the tags to search.
 * @param {number} id the tag id to find.
 * @return {string} the tag.
 */
const getTagById = (tags: Tag[], id: number): Tag => tags
  .find(t => t.id === id) ?? error(`Tag ${id} not found! Corrupted store.`);

type TagsProps = {| +tags: Tag[] |};

/**
 * A function to connect a component with just tags in redux store.
 *
 * @param component the component to connect.
 * @return {ConnectedComponent<Config, TagsProps>} the connected component.
 */
export function tagsConnect<-Config>(
  component: ComponentType<Config>,
): ConnectedComponent<Config, TagsProps> {
  return stateConnect<Config, TagsProps>(
    ({ tags }): TagsProps => ({ tags }),
  )(component);
}

type TagColorProps = {| +getTag: (id: number) => Tag; |};

/**
 * A function to connect a component with a function to find the color of a tag.
 *
 * @param component the component to connect.
 * @return {ConnectedComponent<Config, TagColorProps>} the connected component.
 */
export function getTagConnect<Config: Object>(
  component: ComponentType<Config>,
): ConnectedComponent<Config, TagColorProps> {
  return stateConnect<Config, TagColorProps>(
    ({ tags }): TagColorProps => ({ getTag: id => getTagById(tags, id) }),
  )(component);
}
