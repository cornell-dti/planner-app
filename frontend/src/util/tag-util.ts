import { Tag } from '../store/store-types';

/**
 * ID of the none tag.
 */
export const NONE_TAG_ID: string = 'THE_GLORIOUS_NONE_TAG';

/**
 * The none tag.
 */
export const NONE_TAG: Tag = {
  id: NONE_TAG_ID, order: 0, name: 'None', color: 'gray', classId: null,
};
