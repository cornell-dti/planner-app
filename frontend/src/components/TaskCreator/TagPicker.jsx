// @flow strict

import React from 'react';
import type { Node } from 'react';
import { Icon } from 'semantic-ui-react';
import TagListPicker from '../Util/TagListPicker/TagListPicker';
import styles from './Picker.css';
import { getTagConnect, NONE_TAG_ID } from '../../util/tag-util';
import type { Tag } from '../../store/store-types';

type Props = {|
  +tag: string;
  +opened: boolean;
  +onTagChange: (tag: string) => void;
  +onPickerOpened: () => void;
  // subscribed from redux store.
  +getTag: (id: string) => Tag;
|};

function TagPicker(props: Props): Node {
  const {
    tag, opened, onTagChange, onPickerOpened, getTag,
  } = props;
  // Controllers
  const clickPicker = () => { if (!opened) { onPickerOpened(); } };
  const reset = () => onTagChange(NONE_TAG_ID);
  // Nodes
  const displayedNode = (isDefault: boolean) => {
    const { name, color, classId } = getTag(tag);
    const style = isDefault ? {} : { background: color };
    const internal = isDefault
      ? (<Icon name="tag" className={styles.CenterIcon} />)
      : (
        <React.Fragment>
          <span className={styles.TagDisplay}>{classId == null ? name.split(':')[0] : name}</span>
          <button type="button" className={styles.ResetButton} onClick={reset}>&times;</button>
        </React.Fragment>
      );
    return (
      <span role="presentation" onClick={clickPicker} className={styles.Label} style={style}>
        {internal}
      </span>
    );
  };
  return (
    <div className={styles.Main}>
      {displayedNode(tag === NONE_TAG_ID)}
      {opened && (
        <div className={styles.NewTaskClassPick}>
          <TagListPicker onTagChange={onTagChange} />
        </div>
      )}
    </div>
  );
}

const ConnectedTagPicker = getTagConnect<Props>(TagPicker);
export default ConnectedTagPicker;
