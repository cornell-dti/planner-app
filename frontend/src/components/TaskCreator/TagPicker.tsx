import React, { ReactElement } from 'react';
import { connect } from 'react-redux';
import TagIcon from '../../assets/svgs/tag.svg';
import TagListPicker from '../Util/TagListPicker/TagListPicker';
import styles from './Picker.css';
import { NONE_TAG, NONE_TAG_ID } from '../../util/tag-util';
import { State, Tag } from '../../store/store-types';

type OwnProps = {
  readonly tag: string;
  readonly opened: boolean;
  readonly onTagChange: (tag: string) => void;
  readonly onPickerOpened: () => void;
};
type Props = OwnProps & {
  // subscribed from redux store.
  readonly getTag: (id: string) => Tag;
};

function TagPicker({ tag, opened, onTagChange, onPickerOpened, getTag }: Props): ReactElement {
  // Controllers
  const clickPicker = () => { onPickerOpened(); };
  const reset = () => onTagChange(NONE_TAG_ID);
  // Nodes
  const displayedNode = (isDefault: boolean) => {
    const { name, color, classId } = getTag(tag);
    const style = isDefault ? {} : { background: color };
    const internal = isDefault
      ? (<TagIcon />)
      : (
        <React.Fragment>
          <span className={styles.TagDisplay}>{classId != null ? name.split(':')[0] : name}</span>
          <button type="button" className={styles.ResetButton} onClick={reset}>&times;</button>
        </React.Fragment>
      );
    return (
      <span role="presentation" onClick={clickPicker} className={styles.Label} style={style}>
        {internal}
      </span>
    );
  };
  return (
    <div className={styles.Main}>
      {displayedNode(tag === NONE_TAG_ID)}
      {opened && (
        <div className={styles.NewTaskClassPick}>
          <TagListPicker onTagChange={onTagChange} />
        </div>
      )}
    </div>
  );
}

const Connected = connect(
  ({ tags }: State) => ({ getTag: (id: string) => tags.get(id) || NONE_TAG }),
)(TagPicker);
export default Connected;
