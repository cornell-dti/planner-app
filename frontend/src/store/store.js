// @flow strict

import { createStore } from 'redux';
import type { Store } from 'redux';
import rootReducer from './reducers';
import type { State } from './store-types';
import type { Action } from './action-types';
import type { AppUser } from '../util/firebase-util';
import { error } from '../util/general-util';

export type GlobalStore = Store<State, $Subtype<Action>>;

const store: GlobalStore = createStore(rootReducer);
let appUser: AppUser | null = null;

/**
 * Initialize the app and record a global user and returns the initialized store.
 *
 * @param {AppUser} user the user of the app.
 * @return {GlobalStore} the newly initialized store.
 */
export function initializeApp(user: AppUser): GlobalStore {
  appUser = user;
  return store;
}

/**
 * Returns the global app user.
 *
 * @return {AppUser} the global app user.
 */
export const getAppUser = (): AppUser => appUser ?? error('App is not initialized.');

/**
 * Dispatch an action and returns the dispatched action.
 *
 * @param {Action} action the action to dispatch.
 * @return {Action} the action to dispatch.
 */
export const dispatchAction = (action: Action): Action => store.dispatch(action);
