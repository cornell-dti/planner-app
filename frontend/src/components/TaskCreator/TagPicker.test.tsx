import React from 'react';
import renderer from 'react-test-renderer';
import { ProviderForTesting } from 'store';
import { NONE_TAG_ID } from 'common/lib/util/tag-util';
import TagPicker from './TagPicker';

const dummyHandler = (): void => { };

const createTest = (tag: string, opened: boolean): void => {
  it(`TagPicker(${tag}, ${opened}) matches snapshot`, () => {
    const tree = renderer.create(
      <ProviderForTesting>
        <TagPicker
          tag={tag}
          opened={opened}
          onTagChange={dummyHandler}
          onPickerOpened={dummyHandler}
        />
      </ProviderForTesting>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
};

createTest(NONE_TAG_ID, false);
createTest(NONE_TAG_ID, true);
createTest('foo', false);
createTest('foo', true);
createTest('bar', false);
createTest('bar', true);
createTest('baz', false);
createTest('baz', true);
