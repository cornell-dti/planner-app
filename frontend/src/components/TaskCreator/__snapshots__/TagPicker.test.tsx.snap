// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagPicker(THE_GLORIOUS_NONE_TAG, false) matches snapshot 1`] = `
<div
  className="Main"
>
  <span
    className="Label"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="presentation"
    style={
      Object {
        "background": "#969696",
      }
    }
  >
    <span
      className="TagDisplay"
    >
      <span
        title="Tag"
      >
        <svg
          alt="Tag"
          className="CenterIcon"
          height="1em"
          onKeyUp={[Function]}
          tabIndex={0}
          title="Tag"
          width="1em"
        >
          tag-light.svg
        </svg>
      </span>
       add class  
    </span>
  </span>
</div>
`;

exports[`TagPicker(THE_GLORIOUS_NONE_TAG, true) matches snapshot 1`] = `
<div
  className="Main"
>
  <span
    className="Label"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="presentation"
    style={
      Object {
        "background": "#969696",
      }
    }
  >
    <span
      className="TagDisplay"
    >
      <span
        title="Tag"
      >
        <svg
          alt="Tag"
          className="CenterIcon"
          height="1em"
          onKeyUp={[Function]}
          tabIndex={0}
          title="Tag"
          width="1em"
        >
          tag-light.svg
        </svg>
      </span>
       add class  
    </span>
  </span>
  <div
    className="NewTaskClassPick"
  >
    <ul
      className="NewTaskClass"
    >
      <li
        style={
          Object {
            "--custom-color": "green",
          }
        }
      >
        <input
          onClick={[Function]}
          type="checkbox"
        />
        Baz
      </li>
      <li
        style={
          Object {
            "--custom-color": "blue",
          }
        }
      >
        <input
          onClick={[Function]}
          type="checkbox"
        />
        Bar
      </li>
      <li
        style={
          Object {
            "--custom-color": "red",
          }
        }
      >
        <input
          onClick={[Function]}
          type="checkbox"
        />
        Foo
      </li>
      <li
        style={
          Object {
            "--custom-color": "#969696",
          }
        }
      >
        <input
          onClick={[Function]}
          type="checkbox"
        />
        None
      </li>
    </ul>
  </div>
</div>
`;

exports[`TagPicker(bar, false) matches snapshot 1`] = `
<div
  className="Main"
>
  <span
    className="Label"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="presentation"
    style={
      Object {
        "background": "blue",
      }
    }
  >
    <span
      className="TagDisplay"
    >
      Bar
    </span>
    <button
      className="ResetButton"
      onClick={[Function]}
      type="button"
    >
      ×
    </button>
  </span>
</div>
`;

exports[`TagPicker(bar, true) matches snapshot 1`] = `
<div
  className="Main"
>
  <span
    className="Label"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="presentation"
    style={
      Object {
        "background": "blue",
      }
    }
  >
    <span
      className="TagDisplay"
    >
      Bar
    </span>
    <button
      className="ResetButton"
      onClick={[Function]}
      type="button"
    >
      ×
    </button>
  </span>
  <div
    className="NewTaskClassPick"
  >
    <ul
      className="NewTaskClass"
    >
      <li
        style={
          Object {
            "--custom-color": "green",
          }
        }
      >
        <input
          onClick={[Function]}
          type="checkbox"
        />
        Baz
      </li>
      <li
        style={
          Object {
            "--custom-color": "blue",
          }
        }
      >
        <input
          onClick={[Function]}
          type="checkbox"
        />
        Bar
      </li>
      <li
        style={
          Object {
            "--custom-color": "red",
          }
        }
      >
        <input
          onClick={[Function]}
          type="checkbox"
        />
        Foo
      </li>
      <li
        style={
          Object {
            "--custom-color": "#969696",
          }
        }
      >
        <input
          onClick={[Function]}
          type="checkbox"
        />
        None
      </li>
    </ul>
  </div>
</div>
`;

exports[`TagPicker(baz, false) matches snapshot 1`] = `
<div
  className="Main"
>
  <span
    className="Label"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="presentation"
    style={
      Object {
        "background": "green",
      }
    }
  >
    <span
      className="TagDisplay"
    >
      Baz
    </span>
    <button
      className="ResetButton"
      onClick={[Function]}
      type="button"
    >
      ×
    </button>
  </span>
</div>
`;

exports[`TagPicker(baz, true) matches snapshot 1`] = `
<div
  className="Main"
>
  <span
    className="Label"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="presentation"
    style={
      Object {
        "background": "green",
      }
    }
  >
    <span
      className="TagDisplay"
    >
      Baz
    </span>
    <button
      className="ResetButton"
      onClick={[Function]}
      type="button"
    >
      ×
    </button>
  </span>
  <div
    className="NewTaskClassPick"
  >
    <ul
      className="NewTaskClass"
    >
      <li
        style={
          Object {
            "--custom-color": "green",
          }
        }
      >
        <input
          onClick={[Function]}
          type="checkbox"
        />
        Baz
      </li>
      <li
        style={
          Object {
            "--custom-color": "blue",
          }
        }
      >
        <input
          onClick={[Function]}
          type="checkbox"
        />
        Bar
      </li>
      <li
        style={
          Object {
            "--custom-color": "red",
          }
        }
      >
        <input
          onClick={[Function]}
          type="checkbox"
        />
        Foo
      </li>
      <li
        style={
          Object {
            "--custom-color": "#969696",
          }
        }
      >
        <input
          onClick={[Function]}
          type="checkbox"
        />
        None
      </li>
    </ul>
  </div>
</div>
`;

exports[`TagPicker(foo, false) matches snapshot 1`] = `
<div
  className="Main"
>
  <span
    className="Label"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="presentation"
    style={
      Object {
        "background": "red",
      }
    }
  >
    <span
      className="TagDisplay"
    >
      Foo
    </span>
    <button
      className="ResetButton"
      onClick={[Function]}
      type="button"
    >
      ×
    </button>
  </span>
</div>
`;

exports[`TagPicker(foo, true) matches snapshot 1`] = `
<div
  className="Main"
>
  <span
    className="Label"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="presentation"
    style={
      Object {
        "background": "red",
      }
    }
  >
    <span
      className="TagDisplay"
    >
      Foo
    </span>
    <button
      className="ResetButton"
      onClick={[Function]}
      type="button"
    >
      ×
    </button>
  </span>
  <div
    className="NewTaskClassPick"
  >
    <ul
      className="NewTaskClass"
    >
      <li
        style={
          Object {
            "--custom-color": "green",
          }
        }
      >
        <input
          onClick={[Function]}
          type="checkbox"
        />
        Baz
      </li>
      <li
        style={
          Object {
            "--custom-color": "blue",
          }
        }
      >
        <input
          onClick={[Function]}
          type="checkbox"
        />
        Bar
      </li>
      <li
        style={
          Object {
            "--custom-color": "red",
          }
        }
      >
        <input
          onClick={[Function]}
          type="checkbox"
        />
        Foo
      </li>
      <li
        style={
          Object {
            "--custom-color": "#969696",
          }
        }
      >
        <input
          onClick={[Function]}
          type="checkbox"
        />
        None
      </li>
    </ul>
  </div>
</div>
`;
